### PR TITLE
Use `str::strip_{prefix,suffix}` when possible

### PR DESCRIPTION
**Pull request by tesuji**

### CODE REVIEW

File: src/librustc_typeck/check/intrinsic.rs [hunk 1]
```diff
@@ -436,23 +436,29 @@ pub fn check_platform_intrinsic_type(tcx: TyCtxt<'_>, it: &hir::ForeignItem<'_>)
...
-                    E0439,
-                    "invalid `simd_shuffle`, needs length: `{}`",
-                    name
-                )
-                .emit();
-                return;
+        name if {
+            suffix = name.strip_prefix("simd_shuffle");
+            suffix.is_some()
+        } =>
```
Comment (lzutao) [inline src/librustc_typeck/check/intrinsic.rs:442]: This works around the fact that if-let guard is not implemented (#51114).
At least it gets rid of repeating the pattern in `starts_with` with may lead to typos.

**Comment by Manishearth** [comment]
(I usually don't do arbitrary reviews)

**Comment by lzutao** [comment]
Oh sorry. I thought you were interested to review this PR.

Reassign to random compiler member: r? @petrochenkov

File: src/librustdoc/clean/utils.rs [hunk 2]
```diff
@@ -481,12 +481,14 @@ pub fn print_const(cx: &DocContext<'_>, n: &'tcx ty::Const<'_>) -> String {
...
-                if s.ends_with(": ") {
-                    let n = s.len() - ": ".len();
-                    s.truncate(n);
+            if let Some(head) = s.strip_suffix("_usize") {
+                let new_len = match head.strip_suffix(": ") {
+                    None => head.len(),
+                    Some(hhead) => hhead.len(),
+                };
+                // SAFETY: `new_len` should be in between char boundary
+                unsafe {
```
Comment (Mark-Simulacrum) [inline src/librustdoc/clean/utils.rs:490]: I don't know that the unsafe code here is really worth it -- I can't imagine printing array lengths is going to be very hot, and even if it is truncate() is pretty cheap. Maybe replace with truncate?

File: src/librustdoc/html/markdown.rs [hunk 3]
```diff
@@ -126,26 +145,6 @@ impl<'a> Line<'a> {
     }
 }
 
-// FIXME: There is a minor inconsistency here. For lines that start with ##, we
-// have no easy way of removing a potential single space after the hashes, which
-// is done in the single # case. This inconsistency seems okay, if non-ideal. In
-// order to fix it we'd have to iterate to find the first non-# character, and
-// then reallocate to remove it; which would make us return a String.
```
Comment (Mark-Simulacrum) [inline src/librustdoc/html/markdown.rs:133]: FIXME seems to have gotten lost? I think this isn't fixed.

File: src/librustdoc/html/markdown.rs [hunk 4]
```diff
@@ -111,6 +111,25 @@ enum Line<'a> {
...
+        let trimmed = s.trim();
+        match trimmed.strip_prefix("#") {
+            Some(tail) => match tail.strip_prefix("#") {
+                // `##text` rendered as `#text`.
+                Some(_) => Line::Shown(tail.into()),
+                None => match tail.as_bytes() {
+                    // `#` will be hidden.
+                    [] => Line::Hidden(""),
+                    // `# text` will be hidden.
+                    [b' ' | b'\t', ..] => Line::Hidden(tail),
```
Comment (Mark-Simulacrum) [inline src/librustdoc/html/markdown.rs:124]: Why is tab added here?

File: src/librustdoc/html/markdown.rs [hunk 5]
```diff
@@ -752,9 +753,13 @@ impl LangString {
                     data.ignore = Ignore::All;
                     seen_rust_tags = !seen_other_tags;
                 }
-                x if x.starts_with("ignore-") => {
+                x if {
+                    suffix = x.strip_prefix("ignore-");
+                    suffix.is_some()
```
Comment (Mark-Simulacrum) [inline src/librustdoc/html/markdown.rs:758]: As before I'd prefer to revert this.

File: src/librustdoc/html/markdown.rs [hunk 6]
```diff
@@ -752,9 +753,13 @@ impl LangString {
...
+                } =>
+                {
                     if enable_per_target_ignores {
-                        ignores.push(x.trim_start_matches("ignore-").to_owned());
+                        ignores.push(suffix.unwrap().to_owned());
```
Comment (Mark-Simulacrum) [inline src/librustdoc/html/markdown.rs:762]: This is a behavior change (though probably not in practice); `ignore-ignore-` would get stripped previously.

File: src/librustdoc/html/markdown.rs [hunk 7]
```diff
@@ -775,61 +780,64 @@ impl LangString {
                     seen_rust_tags = !seen_other_tags || seen_rust_tags;
                     data.no_run = true;
                 }
-                x if x.starts_with("edition") => {
-                    data.edition = x[7..].parse::<Edition>().ok();
+                x if {
+                    suffix = x.strip_prefix("edition");
+                    suffix.is_some()
+                } =>
```
Comment (Mark-Simulacrum) [inline src/librustdoc/html/markdown.rs:786]: Please revert this.

File: src/librustdoc/html/markdown.rs [hunk 8]
```diff
@@ -775,61 +780,64 @@ impl LangString {
...
                     } else {
                         seen_other_tags = true;
                     }
                 }
                 x if extra.is_some() => {
+                    let extra = extra.unwrap();
                     let s = x.to_lowercase();
-                    match if s == "compile-fail" || s == "compile_fail" || s == "compilefail" {
-                        Some((
+                    let unknown_error = |flag, help| {
```
Comment (Mark-Simulacrum) [inline src/librustdoc/html/markdown.rs:805]: This looks like an unrelated change?

File: src/librustdoc/html/render.rs [hunk 9]
```diff
@@ -810,11 +810,9 @@ themePicker.onblur = handleThemeButtonsBlur;
                 if line.starts_with(&format!("\"{}\"", krate)) {
                     continue;
                 }
-                if line.ends_with(",\\") {
-                    ret.push(line[..line.len() - 2].to_string());
-                } else {
+                if let Some(head) = line.strip_suffix(",\\").or_else(|| line.strip_suffix("\\")) {
                     // Ends with "\\" (it's the case for the last added crate line)
-                    ret.push(line[..line.len() - 1].to_string());
+                    ret.push(head.to_string());
```
Comment (Mark-Simulacrum) [inline src/librustdoc/html/render.rs:815]: This looks like a change in behavior -- we didn't care if the character was `\` before? Probably benign but I'd prefer to not change behavior at all in this patch.

File: src/librustdoc/passes/collect_intra_doc_links.rs [hunk 10]
```diff
@@ -487,7 +487,7 @@ impl<'a, 'tcx> DocFolder for LinkCollector<'a, 'tcx> {
         }
 
         let cx = self.cx;
-        let dox = item.attrs.collapsed_doc_value().unwrap_or_else(String::new);
+        let dox = item.collapsed_doc_value().unwrap_or_else(String::new);
```
Comment (Mark-Simulacrum) [inline src/librustdoc/passes/collect_intra_doc_links.rs:490]: Unrelated change?

File: src/librustdoc/passes/collect_intra_doc_links.rs [hunk 11]
```diff
@@ -581,28 +574,27 @@ impl<'a, 'tcx> DocFolder for LinkCollector<'a, 'tcx> {
...
-                    link.trim_end_matches("()")
-                } else if link.starts_with("macro@") {
-                    kind = Some(MacroNS);
-                    link.trim_start_matches("macro@")
-                } else if link.starts_with("derive@") {
-                    kind = Some(MacroNS);
-                    link.trim_start_matches("derive@")
-                } else if link.ends_with('!') {
-                    kind = Some(MacroNS);
-                    link.trim_end_matches('!')
```
Comment (Mark-Simulacrum) [inline src/librustdoc/passes/collect_intra_doc_links.rs:601]: These change behavior -- trim_start_matches will strip more than once.

File: src/tools/clippy/clippy_dev/src/ra_setup.rs [hunk 12]
```diff
@@ -51,12 +51,11 @@ fn inject_deps_into_manifest(
...
-        .filter(|line| line.starts_with("extern crate"))
-        // we have something like "extern crate foo;", we only care about the "foo"
-        //              ↓          ↓
-        // extern crate rustc_middle;
-        .map(|s| &s[13..(s.len() - 1)]);
+        // Get the deps
+        //
+        // We have something like "extern crate foo;", we only care about the "foo".
+        // Also we don't trim the string cause we believe rustfmt did that for us.
+        .filter_map(|line| line.strip_prefix("extern crate ").map(|s| s.strip_suffix(';')));
```
Comment (Mark-Simulacrum) [inline src/tools/clippy/clippy_dev/src/ra_setup.rs:58]: This is technically a behavior change -- previously this would work if, say, the character after extern crate is a tab. Regardless I'd personally prefer that the clippy changes go into a PR against the clippy repository rather than here, we should really only be landing changes directly to the subtree if they're "required" (i.e., small patches) to avoid merge conflicts as much as possible.

File: src/tools/compiletest/src/runtest.rs [hunk 13]
```diff
@@ -1192,8 +1192,7 @@ impl<'test> TestCx<'test> {
         for line in reader.lines() {
             match line {
                 Ok(line) => {
-                    let line =
-                        if line.starts_with("//") { line[2..].trim_start() } else { line.as_str() };
+                    let line = line.strip_prefix("//").unwrap_or(&*line).trim_start();
```
Comment (Mark-Simulacrum) [inline src/tools/compiletest/src/runtest.rs:1195]: Behavior change: we only trimmed if stripping previously.

File: src/librustdoc/html/markdown.rs (hunk 4 above)
Comment (ollie27) [reply src/librustdoc/html/markdown.rs:124]: Also the old code would trim the space but this doesn't.

**Comment by lzutao** [comment]
Thanks for the review.
Actually I think I will split this to multiple PRs for easier reviewing.

**Comment by Mark-Simulacrum** [comment]
Having already reviewed most of this (modulo clippy), it would be great for those PRs to incorporate existing feedback at least. Thanks!

File: src/librustc_ast/util/comments.rs [hunk 14]
```diff
@@ -29,27 +29,31 @@ pub struct Comment {
...
+    let yes = match s.as_bytes() {
+        [b'/', b'*', b'*', c, _, ..] => *c != b'*',
+        [b'/', b'*', b'!', _, _, ..] => true,
+        _ => false,
+    };
+    debug!("is {:?} a block doc comment? {}", s, yes);
+    yes
 }
 
-// FIXME(#64197): Try to privatize this again.
```
[the remark this answers is not in the document]
Comment (lzutao) [reply src/librustc_ast/util/comments.rs:47]: it is used here: https://github.com/rust-lang/rust/blob/0c03aee8b81185d65b5821518661c30ecdb42de5/src/librustc_parse/lexer/mod.rs#L185
